components/callbacks/executors.go

```diff
@@ -110,90 +111,56 @@ func (e taskExecutor) executeInvocationTask(
 
-	namespaceTag := metrics.NamespaceTag(ns.Name().String())
-	destTag := metrics.DestinationTag(task.Destination)
-	statusCodeTag := metrics.NexusOutcomeTag(outcomeTag(callCtx, response, callErr))
-	e.MetricsHandler.Counter(RequestCounter.Name()).Record(1, namespaceTag, destTag, statusCodeTag)
-	e.MetricsHandler.Timer(RequestLatencyHistogram.Name()).Record(time.Since(startTime), namespaceTag, destTag, statusCodeTag)
-
-	if callErr == nil {
-		// Body is not read but should be discarded to keep the underlying TCP connection alive.
-		// Just in case something unexpected happens while discarding or closing the body,
-		// propagate errors to the machine.
-		if _, callErr = io.Copy(io.Discard, response.Body); callErr == nil {
-			callErr = response.Body.Close()
-		}
+	saveErr := e.saveResult(callCtx, env, ref, result, err)
+	if saveErr != nil {
+		return saveErr
 	}
 
-	if callErr != nil || response.StatusCode >= 400 {
-		status := "unknown"
-		retryable := callErr != nil
-		if response != nil {
-			status = response.Status
-			retryable = isRetryableHTTPResponse(response)
-		}
-		e.Logger.Error("Callback request failed", tag.Error(callErr), tag.NewStringTag("status", status), tag.NewBoolTag("retryable", retryable))
-	}
-
-	err = e.saveResult(ctx, env, ref, response, callErr)
-
-	if callErr != nil {
-		err = queues.NewDestinationDownError(callErr.Error(), err)
-	} else if isRetryableHTTPResponse(response) {
-		err = queues.NewDestinationDownError(
-			fmt.Sprintf("response returned retryable status code %d", response.StatusCode),
-			err,
-		)
+	// If the request permanently failed there is no need to raise the error
+	if result == Failed {
+		return nil
 	}
 	return err
 }
 
-type invocationArgs struct {
-	url        string
-	header     map[string]string
-	completion nexus.OperationCompletion
-}
-
 func (e taskExecutor) loadInvocationArgs(
 	ctx context.Context,
 	env hsm.Environment,
 	ref hsm.Ref,
-) (args invocationArgs, err error) {
+) (invokable callbackInvokable, err error) {
 	err = env.Access(ctx, ref, hsm.AccessRead, func(node *hsm.Node) error {
 		callback, err := hsm.MachineData[Callback](node)
 		if err != nil {
 			return err
 		}
-		target, err := hsm.MachineData[CanGetNexusCompletion](node.Parent)
-		if err != nil {
-			return err
-		}
+
 		switch variant := callback.GetCallback().GetVariant().(type) {
 		case *persistencespb.Callback_Nexus_:
-			args.url = variant.Nexus.GetUrl()
-			args.header = variant.Nexus.GetHeader()
+			target, err := hsm.MachineData[CanGetNexusCompletion](node.Parent)
+			if err != nil {
+				return err
+			}
+			args := nexusInvocation{}
+			args.nexus = variant.Nexus
 			args.completion, err = target.GetNexusCompletion(ctx)
+			invokable = args
+			if err != nil {
+				return err
+			}
+		case *persistencespb.Callback_Hsm_:
+			target, err := hsm.MachineData[CanGetCompletionEvent](node.Parent)
+			if err != nil {
+				return err
+			}
+			args := hsmInvocation{}
```

Comment:
You don't need `args` here (or above).

```suggestion
			invokable := hsmInvocation{}
```

Reply:
Whoops, commit suggestion didn't work I guess. Fixed.